File: src/api/web/routes/training.ts
import { Router } from 'express';
import { spawn } from 'child_process';
import * as path from 'path';
import { fileURLToPath } from 'url';
import * as fs from 'fs';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const router = Router();

// 训练任务状态管理
interface TrainingTask {
  id: string;
  status: 'running' | 'completed' | 'failed';
  progress: number;
  startTime: string;
  endTime?: string;
  params: TrainingParams;
  result?: any;
  error?: string;
  logs: string[];
}

interface TrainingParams {
  days: number;
  model: 'xgboost' | 'lightgbm' | 'random_forest';
  cvSplits: number;
  useFeatureEngineering: boolean;
}

const trainingTasks = new Map<string, TrainingTask>();

// POST /api/training/start - 启动模型训练
router.post('/start', async (req, res, next) => {
  try {
    const {
      days = 90,
      model = 'xgboost',
      cvSplits = 5,
      useFeatureEngineering = true
    }: TrainingParams = req.body;

    // 验证参数
    if (days < 30 || days > 365) {
      res.status(400);
      throw new Error('days must be between 30 and 365');
    }

    if (!['xgboost', 'lightgbm', 'random_forest'].includes(model)) {
      res.status(400);
      throw new Error('model must be one of: xgboost, lightgbm, random_forest');
    }

    if (cvSplits < 2 || cvSplits > 10) {
      res.status(400);
      throw new Error('cvSplits must be between 2 and 10');
    }

    // 生成任务ID
    const taskId = `train_${Date.now()}`;

    // 创建任务记录
    const task: TrainingTask = {
      id: taskId,
      status: 'running',
      progress: 0,
      startTime: new Date().toISOString(),
      params: { days, model, cvSplits, useFeatureEngineering },
      logs: []
    };
    trainingTasks.set(taskId, task);

    // 构建训练命令
    const scriptPath = path.join(__dirname, '../../../../quant/scripts/ml_retrain.py');
    const args = [
      scriptPath,
      '--days', days.toString(),
      '--model', model,
      '--cv-splits', cvSplits.toString()
    ];

    if (useFeatureEngineering) {
      args.push('--use-feature-engineering');
    }

    // 异步执行训练
    const pythonProcess = spawn('python3', args, {
      cwd: path.join(__dirname, '../../../../quant')
    });

    // 捕获输出
    pythonProcess.stdout.on('data', (data) => {
      const log = data.toString();
      task.logs.push(log);
      console.log(`[Training ${taskId}] ${log}`);

      // 解析进度（如果日志中包含进度信息）
      const progressMatch = log.match(/进度[：:]\s*(\d+)%/);
      if (progressMatch) {
        task.progress = parseInt(progressMatch[1]);
      }
    });

    pythonProcess.stderr.on('data', (data) => {
      const log = data.toString();
      task.logs.push(`ERROR: ${log}`);
      console.error(`[Training ${taskId}] ERROR: ${log}`);
    });

    pythonProcess.on('close', (code) => {
      if (code === 0) {
        task.status = 'completed';
        task.progress = 100;
        task.endTime = new Date().toISOString();

        // 读取训练报告
        try {
          const reportPath = path.join(__dirname, '../../../../quant/quantsys/ml/models/training_report_latest.json');
          if (fs.existsSync(reportPath)) {
            const reportContent = fs.readFileSync(reportPath, 'utf-8');
            task.result = JSON.parse(reportContent);
          }
        } catch (error) {
          console.error('Failed to read training report:', error);
        }

        console.log(`[Training ${taskId}] Completed successfully`);
      } else {
        task.status = 'failed';
        task.endTime = new Date().toISOString();
        task.error = `Training process exited with code ${code}`;
        console.error(`[Training ${taskId}] Failed with code ${code}`);
      }
    });

    res.json({
      success: true,
      data: {
        taskId,
        status: 'running',
        message: 'Training started successfully'
      }
    });
  } catch (error) {
    next(error);
  }
});

// GET /api/training/status/:taskId - 查询训练状态
router.get('/status/:taskId', async (req, res, next) => {
  try {
    const { taskId } = req.params;
    const task = trainingTasks.get(taskId);

    if (!task) {
      res.status(404);
      throw new Error('Training task not found');
    }

    res.json({
      success: true,
      data: {
        id: task.id,
        status: task.status,
        progress: task.progress,
        startTime: task.startTime,
        endTime: task.endTime,
        params: task.params,
        result: task.result,
        error: task.error
      }
    });
  } catch (error) {
    next(error);
  }
});

// GET /api/training/logs/:taskId - 获取训练日志
router.get('/logs/:taskId', async (req, res, next) => {
  try {
    const { taskId } = req.params;
    const task = trainingTasks.get(taskId);

    if (!task) {
      res.status(404);
      throw new Error('Training task not found');
    }

    res.json({
      success: true,
      data: {
        taskId: task.id,
        logs: task.logs
      }
    });
  } catch (error) {
    next(error);
  }
});

// GET /api/training/tasks - 获取所有训练任务
router.get('/tasks', async (req, res, next) => {
  try {
    const tasks = Array.from(trainingTasks.values()).map(task => ({
      id: task.id,
      status: task.status,
      progress: task.progress,
      startTime: task.startTime,
      endTime: task.endTime,
      params: task.params
    }));

    res.json({
      success: true,
      data: tasks
    });
  } catch (error) {
    next(error);
  }
});

// GET /api/training/reports - 获取历史训练报告列表
router.get('/reports', async (req, res, next) => {
  try {
    const modelsDir = path.join(__dirname, '../../../../quant/quantsys/ml/models');

    if (!fs.existsSync(modelsDir)) {
      res.json({ success: true, data: [] });
      return;
    }

    const files = fs.readdirSync(modelsDir);
    const reportFiles = files
      .filter(f => f.startsWith('training_report_') && f.endsWith('.json') && f !== 'training_report_latest.json')
      .sort()
      .reverse();

    const reports = reportFiles.slice(0, 20).map(filename => {
      const filePath = path.join(modelsDir, filename);
      const content = fs.readFileSync(filePath, 'utf-8');
      const report = JSON.parse(content);

      // 从文件名提取时间戳
      const timestamp = filename.replace('training_report_', '').replace('.json', '');

      return {
        filename,
        timestamp,
        metrics: report.metrics,
        params: report.params,
        n_features: report.n_features
      };
    });

    res.json({
      success: true,
      data: reports
    });
  } catch (error) {
    next(error);
  }
});

// GET /api/training/report/:filename - 获取特定训练报告详情
router.get('/report/:filename', async (req, res, next) => {
  try {
    const { filename } = req.params;

    // 安全检查：防止路径遍历攻击
    if (filename.includes('..') || filename.includes('/')) {
      res.status(400);
      throw new Error('Invalid filename');
    }

    const reportPath = path.join(__dirname, '../../../../quant/quantsys/ml/models', filename);

    if (!fs.existsSync(reportPath)) {
      res.status(404);
      throw new Error('Report not found');
    }

    const content = fs.readFileSync(reportPath, 'utf-8');
    const report = JSON.parse(content);

    res.json({
      success: true,
      data: report
    });
  } catch (error) {
    next(error);
  }
});

// GET /api/training/history - 获取训练历史记录
router.get('/history', async (req, res, next) => {
  try {
    const modelsDir = path.join(__dirname, '../../../../quant/quantsys/ml/models');

    if (!fs.existsSync(modelsDir)) {
      res.json({ history: [] });
      return;
    }

    const files = fs.readdirSync(modelsDir);
    const reportFiles = files
      .filter(f => f.startsWith('training_report_') && f.endsWith('.json') && f !== 'training_report_latest.json')
      .sort()
      .reverse();

    const history = reportFiles.slice(0, 50).map(filename => {
      try {
        const filePath = path.join(modelsDir, filename);
        const content = fs.readFileSync(filePath, 'utf-8');
        const report = JSON.parse(content);

        return {
          timestamp: report.timestamp,
          model_type: report.model_type,
          n_features: report.data?.n_features || 0,
          total_samples: report.data?.total_samples || 0,
          cv_accuracy: report.cv_results?.mean_scores?.accuracy || 0,
          cv_auc: report.cv_results?.mean_scores?.auc || 0,
          test_accuracy: report.test_metrics?.accuracy || 0,
          test_auc: report.test_metrics?.auc || 0,
          class_balance: report.data?.class_balance || 0
        };
      } catch (error) {
        console.warn(`Failed to parse training report ${filename}:`, error);
        return null;
      }
    }).filter(record => record !== null);

    res.json({ history });
  } catch (error) {
    next(error);
  }
});

export { router as trainingRouter };

File: quantsys-v2/adapters/inbound/api/routes/training.py
"""
training routes.
"""
import json
import sys
from datetime import datetime, timedelta
from pathlib import Path
import re
import uuid

from flask import Blueprint, jsonify, request

import os

from adapters.inbound.api.shared import (
    ds,
    api_response,
    handle_api_error,
    sanitize_for_json,
    convert_keys_to_snake,
    convert_keys_to_camel,
    _safe_float,
    _V2_ROOT,
    _PROJECT_ROOT_PATH,
    _LEGACY_QUANT_ROOT,
    _load_pipeline_runs,
    _save_pipeline_runs,
    _get_pipeline_run,
    _update_pipeline_run,
    acquire_task,
    release_task,
    get_running_tasks_snapshot,
    strategy_service,
    stock_pool_service,
    factor_adapter,
    scoring_service,
    _read_watchlist,
    _write_watchlist,
    _read_groups,
    _write_groups,
    _parse_sina_a_quote,
    _parse_sina_hk_quote,
    to_camel_case,
    to_snake_case,
    get_query_params_snake_case,
    enrich_stock_data,
    signal_to_opportunity,
)

training_bp = Blueprint('training', __name__)

_TRAINING_REPORTS_DIR = Path(os.getcwd()) / 'ml' / 'models'


@training_bp.route('/api/training/start', methods=['POST'])
@handle_api_error
def training_start():
    """启动模型训练（兼容 Express 前端）"""
    try:
        from application.services.ml_pipeline.trainer import MLTrainer
        from application.services.ml_pipeline.feature_engineering import FeatureEngineer
        from datetime import datetime, timedelta
        import pandas as pd
    except ImportError as e:
        return jsonify({'success': False, 'error': f'ML module not available: {e}'}), 503

    data = request.get_json(silent=True) or {}

    model_type = data.get('model', data.get('model_type', 'xgboost'))
    days = data.get('days', 90)
    cv_splits = data.get('cvSplits', data.get('cv_splits', 5))
    start_date = data.get('start_date') or (datetime.now() - timedelta(days=days)).strftime('%Y-%m-%d')
    end_date = data.get('end_date') or datetime.now().strftime('%Y-%m-%d')
    test_size = 1.0 / max(cv_splits, 1)

    all_stocks = ds.stock.get_all(limit=1000)
    stock_list = [s['symbol'] for s in all_stocks[:100]]

    klines_dict = {}
    for symbol in stock_list:
        klines_df = ds.kline.get_daily_klines(symbol, start_date, end_date)
        if klines_df is not None and not klines_df.is_empty() and len(klines_df) >= 120:
            klines_dict[symbol] = klines_df.to_dicts()

    if len(klines_dict) == 0:
        return jsonify({'success': False, 'error': 'No sufficient training data available'}), 400

    engineer = FeatureEngineer(scaler_type='standard')
    features_df = engineer.extract_features(klines_dict)

    if 'roc_5' in features_df.columns:
        target = (features_df['roc_5'].fillna(0) > 0).astype(int)
    else:
        target = (features_df.select_dtypes(include=['float64', 'int64']).iloc[:, 0].fillna(0) > 0).astype(int)

    metadata, scaled_features = engineer.prepare_features(
        features_df, handle_missing='drop', fit_scaler=True
    )
    target = target[scaled_features.index]

    if len(scaled_features) < 50:
        return jsonify({'success': False, 'error': 'Insufficient training samples after preprocessing'}), 400

    trainer = MLTrainer(model_type=model_type)
    training_results = trainer.train(scaled_features, target, test_size=test_size, params=data.get('params'))
    model_path = trainer.save_model(version='latest')

    report = {
        'timestamp': datetime.now().isoformat(),
        'start_time': datetime.now().isoformat(),
        'end_time': datetime.now().isoformat(),
        'duration_seconds': 0,
        'model_type': model_type,
        'n_features': int(scaled_features.shape[1]) if hasattr(scaled_features, 'shape') else 0,
        'total_samples': int(len(scaled_features)),
        'metrics': training_results.get('test_metrics', {}),
        'cv_results': training_results.get('cv_results', {}),
        'params': data,
    }
    _TRAINING_REPORTS_DIR.mkdir(parents=True, exist_ok=True)
    report_filename = f"training_report_{datetime.now().strftime('%Y%m%d_%H%M%S')}.json"
    report_path = _TRAINING_REPORTS_DIR / report_filename
    report_path.write_text(json.dumps(sanitize_for_json(report), ensure_ascii=False, indent=2), encoding='utf-8')

    latest_path = _TRAINING_REPORTS_DIR / 'training_report_latest.json'
    latest_path.write_text(json.dumps(sanitize_for_json(report), ensure_ascii=False, indent=2), encoding='utf-8')

    return api_response({
        'model_path': str(model_path),
        'training_results': sanitize_for_json(training_results),
        'samples_trained': len(scaled_features),
        'symbols_count': len(klines_dict),
        'report_filename': report_filename,
    }, message='Training completed')


@training_bp.route('/api/training/reports', methods=['GET'])
@handle_api_error
def training_reports():
    """列出训练报告"""
    if not _TRAINING_REPORTS_DIR.exists():
        return api_response({'reports': []})

    files = sorted(
        [f for f in _TRAINING_REPORTS_DIR.iterdir()
         if f.name.startswith('training_report_') and f.name.endswith('.json')
         and f.name != 'training_report_latest.json'],
        reverse=True
    )
    reports = []
    for fp in files[:20]:
        try:
            content = json.loads(fp.read_text(encoding='utf-8'))
            reports.append({
                'filename': fp.name,
                'timestamp': content.get('timestamp', ''),
                'metrics': content.get('metrics', {}),
                'params': content.get('params', {}),
                'n_features': content.get('n_features', 0),
                'total_samples': content.get('total_samples', 0),
                'model_type': content.get('model_type', ''),
            })
        except (json.JSONDecodeError, OSError):
            pass

    return api_response({'reports': reports, 'count': len(reports)})


@training_bp.route('/api/training/report/<filename>', methods=['GET'])
@handle_api_error
def training_report_detail(filename):
    """获取指定训练报告"""
    if '..' in filename or '/' in filename or '\\' in filename:
        return jsonify({'success': False, 'error': 'Invalid filename'}), 400

    report_path = _TRAINING_REPORTS_DIR / filename
    if not report_path.exists():
        return jsonify({'success': False, 'error': 'Report not found'}), 404

    try:
        content = json.loads(report_path.read_text(encoding='utf-8'))
        return api_response(content)
    except json.JSONDecodeError:
        return jsonify({'success': False, 'error': 'Invalid report file'}), 500


@training_bp.route('/api/training/history', methods=['GET'])
@handle_api_error
def training_history():
    """训练历史摘要"""
    if not _TRAINING_REPORTS_DIR.exists():
        return api_response({'history': [], 'count': 0})

    files = sorted(
        [f for f in _TRAINING_REPORTS_DIR.iterdir()
         if f.name.startswith('training_report_') and f.name.endswith('.json')
         and f.name != 'training_report_latest.json'],
        reverse=True
    )
    history = []
    for fp in files[:50]:
        try:
            content = json.loads(fp.read_text(encoding='utf-8'))
            cv_results = content.get('cv_results', {}) or {}
            metrics = content.get('metrics', {}) or {}
            history.append({
                'timestamp': content.get('timestamp', ''),
                'start_time': content.get('start_time', ''),
                'end_time': content.get('end_time', ''),
                'duration_seconds': content.get('duration_seconds', 0),
                'model_type': content.get('model_type', ''),
                'n_features': content.get('n_features', 0),
                'total_samples': content.get('total_samples', 0),
                'cv_accuracy': cv_results.get('mean_scores', {}).get('accuracy', 0),
                'cv_auc': cv_results.get('mean_scores', {}).get('auc', 0),
                'test_accuracy': metrics.get('accuracy', 0),
                'test_auc': metrics.get('auc', 0),
            })
        except (json.JSONDecodeError, OSError):
            pass

    return api_response({'history': history, 'count': len(history)})

